gitutil.py

```diff
@@ -0,0 +1,143 @@
+    ensure_not_empty(email)
+
+    repo = git.Repo(repo_path)
+    _ensure_submodule_exists(repo, submodule_path)
+
+    # Create mk-tree-parseable string-representation from given tree-ish.
+    tree = repo.tree(tree_ish)
+    tree_representation = _serialise_and_patch_tree(tree, submodule_path, commit_hash)
+
+    # Pass the patched tree to git mk-tree using GitPython. We cannot do this in GitPython
+    # directly as it does not support arbitrary tree manipulation.
+    # We must keep a reference to auto_interrupt as it closes all streams to the subprocess
+    # on finalisation
+    auto_interrupt = repo.git.mktree(istream = subprocess.PIPE, as_process=True)
+    process = auto_interrupt.proc
+    stdout, _= process.communicate(input=tree_representation.encode())
+
+    # returned string is byte-encoded and newline-terminated
+    new_sha = stdout.decode('utf-8').strip()
+
+    # Create a new commit in the repo's object database from the newly created tree.
+    actor = git.Actor(author, email)
+    parent_commit = repo.commit(tree_ish)
+    commit = git.Commit.create_from_tree(
+      repo = repo,
+      tree = new_sha,
+      parent_commits = [parent_commit],
+      message='Upgrade submodule {s} to commit {c}'.format(
+          s=submodule_path,
+          c=commit_hash,
+      ),
+      author=actor,
+      committer=actor,
+    )
+
+    return commit.hexsha
+
+
+def _serialise_and_patch_tree(
+    tree: git.Tree,
+    submodule_path: str,
+    commit_hash: str,
+):
+    '''Return a modified, serialised tree-representation in which the given submodule's entry is altered.
+    The resulting serialisation can be parsed by git mk-tree to create a new tree-object.
+    '''
+    # Since there is no direct way to get the ls-tree representation from GitPython we need to gather
+    # the necessary information for each element of the given tree ourselves.
+    serialised_tree = ""
+    for tree_element in tree:
+        # GitPython uses the special type 'submodule' for submodules whereas git uses 'commit'.
+        if tree_element.type == 'submodule':
+            element_type = 'commit'
+            # Replace the hash the of the 'commit'-tree with the passed value if the submodule is at the specified path
+            if tree_element.path == submodule_path:
+                element_sha = commit_hash
+                submodule_found = True
+        else:
+            element_type = tree_element.type
+            element_sha = tree_element.hexsha
+
+        serialised_tree+='{mode} {type} {sha}\t{path}\n'.format(
+            sha=element_sha,
+            type=element_type,
+            # mode is a number in octal representation WITHOUT '0o' prefix
+            mode=format(tree_element.mode, 'o'),
+            path=tree_element.path,
+        )
+    return serialised_tree.strip('\n')
+
+
+def _ensure_submodule_exists(repo: git.Repo, path: str):
+    '''Use GitPython to verify that a submodule with the given path exists in the repository.'''
```

Comment:
it is irrelevant which library we are using; in particluar, this can be seen in the code